src/Http/Routing/src/RouteEndpointDataSource.cs

```diff
@@ -0,0 +1,262 @@
+        var handler = entry.RouteHandler;
+        var displayName = pattern.RawText ?? pattern.DebuggerToString();
+
+        // Methods defined in a top-level program are generated as statics so the delegate target will be null.
+        // Inline lambdas are compiler generated method so they be filtered that way.
+        if (GeneratedNameParser.TryParseLocalFunctionName(handler.Method.Name, out var endpointName)
+            || !TypeHelper.IsCompilerGeneratedMethod(handler.Method))
+        {
+            endpointName ??= handler.Method.Name;
+            displayName = $"{displayName} => {endpointName}";
+        }
+
+        if (entry.HttpMethods is not null)
+        {
+            // Prepends the HTTP method to the DisplayName produced with pattern + method name
+            displayName = $"HTTP: {string.Join(", ", entry.HttpMethods)} {displayName}";
+        }
+
+        if (entry.IsFallback)
+        {
+            displayName = $"Fallback {displayName}";
+        }
+
+        RequestDelegate? factoryCreatedRequestDelegate = null;
+        RequestDelegate redirectedRequestDelegate = context =>
+        {
+            if (factoryCreatedRequestDelegate is null)
+            {
+                throw new InvalidOperationException(Resources.RouteEndpointDataSource_RequestDelegateCannotBeCalledBeforeBuild);
+            }
+
+            return factoryCreatedRequestDelegate(context);
+        };
+
+        // The Map methods don't support customizing the order apart from using int.MaxValue to give MapFallback the lowest priority.
+        // Otherwise, we always use the default of 0 unless a convention changes it later.
+        var order = entry.IsFallback ? int.MaxValue : 0;
+
+        RouteEndpointBuilder builder = new(redirectedRequestDelegate, pattern, order)
+        {
+            DisplayName = displayName,
+            ApplicationServices = _applicationServices,
+        };
+
+        // We own EndpointBuilder.Metadata (in another assembly), so we know it's just a List.
+        var metadata = (List<object>)builder.Metadata;
+
+        // Add MethodInfo as first metadata item
+        builder.Metadata.Add(handler.Method);
+
+        if (entry.HttpMethods is not null)
+        {
+            builder.Metadata.Add(new HttpMethodMetadata(entry.HttpMethods));
+        }
+
+        // Apply group conventions before entry-specific conventions added to the RouteHandlerBuilder.
+        if (groupConventions is not null)
+        {
+            foreach (var groupConvention in groupConventions)
+            {
+                groupConvention(builder);
+            }
+        }
+
+        // Add delegate attributes as metadata before entry-specific conventions but after group conventions.
+        var attributes = handler.Method.GetCustomAttributes();
+        if (attributes is not null)
+        {
+            metadata.AddRange(attributes);
+        }
+
+        entry.Conventions.HasBeenBuilt = true;
```

Comment:
I was confused by HasBeenBuilt. Where is the call to Build? Should this be `IsReadOnly` instead? Also add a comment about why this place in the code is where you can no longer add conventions (because we're going to be using them now and new ones won't be observed etc)